File: key/key.go
package key

import (
	"bytes"
	"crypto/rand"
	"crypto/subtle"
	"encoding/base64"
	"errors"
	"fmt"
	"strings"

	"github.com/google/uuid"
	"golang.org/x/crypto/curve25519"
)

// KeySize is Default curve25519 key size
// const RawKeyLen = 32

type RawKey [32]byte

func AsRawKey(my []byte) *RawKey {
	ret := RawKey{}
	if len(my) < len(ret) {
		for i := range my {
			ret[i] = my[i]
		}
	} else {
		for i := range ret {
			ret[i] = my[i]
		}
	}
	return &ret
}

// Uncool to copie a key in local memory
func (my *RawKey) As32Byte() *[32]byte {
	out := [32]byte{}
	for i := range *my {
		out[i] = my[i]
	}
	return &out
}

type KeyStyle string

const (
	Private = KeyStyle("Private")
	Public  = KeyStyle("Public")
)

type KeyType struct {
	Id    string
	Style KeyStyle
	Raw   RawKey
}

type JsonKeyType struct {
	Id    string
	Style KeyStyle
	Raw   string
}

func toKey(i []byte) RawKey {
	ret := RawKey{}
	if len(i) < len(ret) {
		for k := range i {
			ret[k] = i[k]
		}
	} else {
		for k := range ret {
			ret[k] = i[k]
		}
	}
	return ret
}

func (k *KeyType) AsJson() *JsonKeyType {
	return &JsonKeyType{
		Id:    k.Id,
		Style: k.Style,
		Raw:   base64.StdEncoding.EncodeToString([]byte(k.Raw[:])),
	}
}

func ToKeyType(jsk JsonKeyType) (*KeyType, error) {
	raw, err := base64.StdEncoding.DecodeString(jsk.Raw)
	if err != nil {
		return nil, err
	}
	return &KeyType{
		Id:    jsk.Id,
		Style: jsk.Style,
		Raw:   toKey(raw),
	}, nil
}

type PublicKey struct {
	Key KeyType
}

type PrivateKey struct {
	Key KeyType
}

func MakePublicKey(key *RawKey, id string) *PublicKey {
	return &PublicKey{
		KeyType{
			Id:    id,
			Style: Public,
			Raw:   *key,
		},
	}
}
func MakePrivateKey(key *RawKey, ids ...string) *PrivateKey {
	var id string
	if len(ids) == 0 || (len(ids) > 0 && len(ids[0]) == 0) {
		id = uuid.New().String()
	} else {
		id = ids[0]
	}
	return &PrivateKey{
		KeyType{
			Id:    id,
			Style: Private,
			Raw:   *key,
		},
	}
}

// CreateRandomKey generates a new random key.
func CreateRandomKey() (*RawKey, error) {
	k := RawKey{}
	_, err := rand.Read(k[:])
	if err != nil {
		return nil, err
	}
	return &k, nil
}

// NewPrivateKey generates a new curve25519 secret key.
// It conforms to the format described on https://cr.yp.to/ecdh.html.
func NewPrivateKey(pk ...*PrivateKey) (*PrivateKey, error) {
	var k RawKey
	if len(pk) > 0 && pk[0] != nil &&
		pk[0].Key.Style == Private &&
		len(pk[0].Key.Raw) == len(RawKey{}) {
		k = pk[0].Key.Raw
	} else {
		m, err := CreateRandomKey()
		if err != nil {
			return &PrivateKey{}, err
		}
		k = *m
	}
	k[0] &= 248
	k[31] = (k[31] & 127) | 64
	return MakePrivateKey(&k), nil
}

func isZero(k RawKey) bool {
	zeros := RawKey{}
	return subtle.ConstantTimeCompare(zeros[:], k[:]) == 1
}

// Public computes the public key matching this curve25519 secret key.
func (k *PrivateKey) Public() *PublicKey {
	if isZero(k.Key.Raw) {
		panic("Tried to generate emptyPrivateKey.Public()")
	}
	pub := [len(RawKey{})]byte{}
	curve25519.ScalarBaseMult(&pub, k.Key.Raw.As32Byte())
	return MakePublicKey(AsRawKey(pub[:]), k.Key.Id)
}

// MarshalText create a String of the Private Key
func (k *PrivateKey) Marshal() string {
	buf := new(bytes.Buffer)
	fmt.Fprintf(buf, `privkey:%s`, k.Key.Raw.Marshal())
	return buf.String()
}

func (k *RawKey) Marshal() string {
	return base64.StdEncoding.EncodeToString(k[:])
}

func (k *PublicKey) Marshal() string {
	return k.Key.Raw.Marshal()
}

func fromText(pkstr string) (*RawKey, error) {
	ret := RawKey{}
	dc, err := base64.StdEncoding.DecodeString(pkstr)
	if err != nil {
		return nil, err
	}
	if len(dc) != len(RawKey{}) {
		return nil, errors.New("this is not a rawkey")
	}
	for i := range dc {
		ret[i] = dc[i]
	}
	return &ret, nil
}

func FromText(pkstr string, id string) (*PrivateKey, *PublicKey, error) {
	if strings.HasPrefix(pkstr, "privkey:") {
		keyBytes, err := fromText(strings.TrimPrefix(pkstr, "privkey:"))
		if err != nil {
			return nil, nil, err
		}
		return MakePrivateKey(keyBytes, id), nil, nil
	}
	keyBytes, err := fromText(pkstr)
	if err != nil {
		return nil, nil, err
	}
	return nil, MakePublicKey(keyBytes, id), nil
}

File: key/key.ts
import * as uuid from 'uuid';
import { Fault, Ok, Result } from '../utils/render';
// package key

// import (
// 	"bytes"
// 	"crypto/rand"
// 	"crypto/subtle"
// 	"encoding/base64"
// 	"errors"
// 	"fmt"
// 	"strings"

// 	"github.com/google/uuid"
// 	"golang.org/x/crypto/curve25519"
// )

// KeySize is Default curve25519 key size
const RawKeyLen = 32;

export class RawKey {
	public readonly bytes = new Uint8Array(RawKeyLen)
	constructor(my: Uint8Array | Buffer) {
		const ret = this.bytes
		if (my.length < ret.length) {
			for (let i = 0; i < my.length; ++i) {
				ret[i] = my[i]
			}
		} else {
			for (let i = 0; i < ret.length; ++i) {
				ret[i] = my[i]
			}
		}
	}
	public As32Byte(): Uint8Array {
		return this.bytes
	}
	public Marshal(): string {
		const out = Buffer.from(this.bytes);
		return out.toString('base64')
	}

}

export enum KeyStyle {
	Private = "Private",
	Public  = "Public"
}

export interface JsonKeyType {
	Id:    string
	Style: KeyStyle
	Raw:   string
}

export function ToKeyType(jsk: JsonKeyType): KeyType | undefined {
	const raw = Buffer.from(jsk.Raw, 'base64')
	return new KeyType(jsk.Id, jsk.Style, raw);
}


export class KeyType {
	public readonly Raw:   RawKey
	constructor(
	public readonly Id:    string,
	public readonly Style: KeyStyle,
	key: Uint8Array|Buffer
	) {
		this.Raw = new RawKey(key)
	}

	public AsJSON(): JsonKeyType {
		let buff = Buffer.from(this.Raw.As32Byte());
		return {
			Id:    this.Id,
			Style: this.Style,
			Raw:   buff.toString('base64')
		}
	}
}



export class PublicKey extends KeyType {
	readonly Style = KeyStyle.Public
	public Marshal(): string {
		return this.Raw.Marshal()
	}

}

export class PrivateKey extends KeyType {
	readonly Style = KeyStyle.Private
	// MarshalText create a String of the Private Key
	public Marshal(): string {
		return `privkey:${this.Raw.Marshal()}`
	}
	// Public computes the public key matching this curve25519 secret key.
	public Public(): PublicKey {
	if (isZero(this.Raw)) {
		throw Error("Tried to generate emptyPrivateKey.Public()")
	}
	const pub = new Uint8Array(RawKeyLen)
	throw Error("not ready")
	// curve25519.ScalarBaseMult(&pub, k.Key.Raw.As32Byte())
	return MakePublicKey(pub, this.Id)
}

}

export function MakePublicKey(key: Uint8Array, id: string): PublicKey {
	return new PublicKey( id, KeyStyle.Public, key)
}

export function MakePrivateKey(key :Uint8Array, id?: string) :PrivateKey {
	return new PrivateKey(id || uuid.v4(), KeyStyle.Private, key)
}

// CreateRandomKey generates a new random key.
export function CreateRandomKey(): Result<RawKey> {
	return Fault("missing implementation")
	// return new RawKey()
}

// NewPrivateKey generates a new curve25519 secret key.
// It conforms to the format described on https://cr.yp.to/ecdh.html.
export function NewPrivateKey(pk?: PrivateKey, id?: string): PrivateKey {
	let k: Uint8Array
	if (pk && pk.Raw.bytes[0] &&
		pk.Style == KeyStyle.Private &&
		pk.Raw.bytes.length === RawKeyLen) {
		k = pk.Raw.bytes
	} else {
		k = CreateRandomKey().val.bytes
	}
	k[0] &= 248
	k[31] = (k[31] & 127) | 64
	return MakePrivateKey(k, id)
}

export function isZero(k: RawKey): boolean {
	return !k.bytes.find(i => i !== 0)
}



export function fromText(pkstr: string): RawKey {
	return new RawKey(Buffer.from(pkstr, 'base64'))
}

export interface FromTextResult {
	PrivateKey?: PrivateKey,
	PublicKey?: PublicKey,
}
export function FromText(pkstr: string, id: string): Result<FromTextResult> {
	if (pkstr.startsWith("privkey:")) {
		const keyBytes = fromText(pkstr.substr("privkey:".length))
		return Ok({
			PrivateKey: MakePrivateKey(keyBytes.bytes, id)
		})
	}
	const keyBytes = fromText(pkstr)
	return Ok({
		PublicKey: MakePublicKey(keyBytes.bytes, id)
	})
}
